### DISCUSSION TITLE
Ctrl R not working after upgrading iterm2 to ohmysh

### DISCUSSION THREAD

**Discussion by tmel173** [ohmyzsh/ohmyzsh#11893] [Help] [answered]
Sometime ago installed iterm2 as a I usually do on all my Macs terminal but after I went to ohmzsh I seem to have lost my "ctrl + r" cycling through my history for some reason.

I resolved this issue on my work computer which was also installed iTerm2 but no OHMYZSH. I resolved it by remapping the keybindings in the OHMZSH config file ~/.zshrc as below :

```
#source /usr/share/doc/fzf/examples/key-bindings.zsh
bindkey -M viins 'ˆr' history-incremental-search-backward
bindkey -M vicmd 'ˆr' history-incremental-search-backward
bindkey -v
```

This didn't work on my personal laptop and can't seem to figure out what the issue is. I could try remove ohmyzsh but I'm reluctant to do that as I had customized so beautifully.

**Comment by carlosala** [comment, marked as answer]
If you want to use `vi` keybindings, I recommend using the `vi-mode`plugin and not changing anything else about `bindkey`.
`^r` will work out of the box.

**Reply by tmel173** [reply]
please forgive me but I'm a little lost with your response☺️It seems that you are saying the history cycling will be recovered from installing the vi-mode plugin but I'm not fully comprehending the logic. This is what I mean - every linux terminal comes with the function [ctrl +r] to cycle through your commands history. In my case I have iTerm2 as my custom terminal as mentioned and that works without adding ohmyzsh. I cant understand the correlation between vi-mode plugin fixing that.

Please can you be more clear how that works ?

**Reply by tmel173** [reply]
This resolved my issue btw but I wanted to understand further. Thanks very much!

**Reply by carlosala** [reply]
@tmel173 of course, no problem! By default, zsh does not have the `^r` shortcut for reverse history search (you can reproduce it by executing `zsh -f` and trying!). We add it [here](https://github.com/ohmyzsh/ohmyzsh/blob/bbda81fe4b338f00bbf7c7f33e6d1b12d067dc05/lib/key-bindings.zsh#L109), but if you run `bindkey -v` after loading `oh-my-zsh.sh` you override all key-bindings. You can reproduce this by executing in a clean ohmyzsh installation (with no `bindkey` calls):
```zsh
bindkey | grep '\^Xr' # you'll see an entry
bindkey -v # set viins keymap
bindkey | grep '\^Xr' # no entry anymore
```
By adding the `vi-mode` plugin, you get the `bindkey -v` executed but maintain all the good stuff like the `^r` keymap.
I hope this explanation fulfils your doubts!

**Reply by tmel173** [reply]
That explains why other shortcut keys were also broken ! Thanks a lot for the explanation !

**Reply by rwmitchell** [reply]
more importantly, OMZ sets bindkey to emacs mode in the startup code, overriding anything you may have already set.  vi-mode sets it back to vi.  zsh already defaults to emacs mode, OMZ should not be setting it in the startup.

**Reply by tmel173** [reply]
not sure what happened but I am facing a new issue with something a little interesting where my <ctrl + r> only works as root. I do notice thats also affecting my ssh only working with root on remote servers. I don't mind that later part working for that part but its the <ctrl + r> bothering me. I was thinking of opening a new ticket but it seems tied to this issue. I'm happy to open a new ticket if that helps
